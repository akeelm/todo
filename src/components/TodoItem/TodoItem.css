.todo-item {

}

.todo-item__text--completed {
    text-decoration: line-through;
}