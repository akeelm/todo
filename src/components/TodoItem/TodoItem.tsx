import React from 'react';

export interface ITodoItem {
    id: number;
    text: string;
    completed: boolean;
}

const TodoItem: React.FC<ITodoItem> = ({
    id,
    text,
    completed
}) => {
    return (
        <div className="todo-item" data-testid="todo-item">

        </div>
    );
}

export default TodoItem;
