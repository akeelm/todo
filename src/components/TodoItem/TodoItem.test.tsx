import { render, getByTestId, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import TodoItem from '.';
import { ITodoItemProps } from './TodoItem'

describe('TodoItem tests', () => {
    const mockCompleteHandler = (id: number, text: string, completed: boolean) => jest.fn();
    const mockDeleteHandler = (id: number) => jest.fn();
    const mockEditFn = jest.fn()
    const mockEditHandler = (id: number, string: string, completed: boolean) => mockEditFn();
    const mockTodoItem: ITodoItemProps = {
        id: 1,
        text: 'Need to write some tests',
        completed: false,
        completeHandler: mockCompleteHandler,
        deleteHandler: mockDeleteHandler,
        editHandler: mockEditHandler
    };

   it('should be able to render', () => {
       const { container } = render(
           <TodoItem {...mockTodoItem} />
       );

       expect(container).toMatchSnapshot();
   });

   it ('should have an editable input', async () => {
       const { container } = render(<TodoItem {...mockTodoItem} />);
       const text: Element = getByTestId(container, 'todo-item-text');
       await userEvent.click(text);

       const input: Element = getByTestId(container, 'todo-item-input');
       await waitFor(() => expect(input).toBeTruthy());

       await userEvent.type(input, 'test input{enter}');
       expect(mockEditFn).toHaveBeenCalled();
   });

   it ('should be able to mark as completed', async() => {
       const { container } = render(
           <TodoItem {...mockTodoItem} completed={true} />
       );

       const text: Element = getByTestId(container, 'todo-item-text');
       expect(text).toHaveClass('todo-item__text--completed');
   });

   it('should be able to delete the todo', () => {
       const { container } = render(
           <TodoItem {...mockTodoItem} />
       );

       const deleteButton: Element = getByTestId(container, 'delete-todo-button');
       userEvent.click(deleteButton);
       waitFor(() => expect(mockDeleteHandler).toHaveBeenCalled());
   });

   it('shouldnt have delete or complete checkbox if new todo', () => {
       const { queryByTestId } = render(
           <TodoItem {...mockTodoItem} id={-1} />
       );

       expect(queryByTestId ('delete-todo-button')).toBeNull();
       expect(queryByTestId ('todo-item-checkbox')).toBeNull();
   })
});