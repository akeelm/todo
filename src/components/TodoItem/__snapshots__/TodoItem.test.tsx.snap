// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem tests should be able to render 1`] = `
<div>
  <li
    class="todo-item"
    data-testid="todo-item"
  >
    <input
      data-testid="todo-item-checkbox"
      tabindex="0"
      title="Complete"
      type="checkbox"
    />
    <span
      class="todo-item__text "
      data-testid="todo-item-text"
      tabindex="0"
    >
      Need to write some tests
    </span>
    <button
      data-testid="delete-todo-button"
      title="Delete todo"
    >
      Delete
    </button>
  </li>
</div>
`;
